### PR TITLE
`Exam mode`: Collect exam interactions and setup initial monitoring structure

### PR DESCRIPTION
**Pull request by Dominik-Weinzierl**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

As part of my bachelor thesis, we want to get more insight into the exam. Especially as an instructor it is difficult to understand in retrospect how the exam went and especially how the students dealt with it. In order to avoid this problem, we offer the possibility to monitor the exam. This pull request provides the basic structure for collecting the user exam interactions on the client side and persisting them on the server side.

### Description

Each action is collected on the client and stored as part of the `StudentExam`.
To synchronize the actions with the server, we use the same logic which we use to submit and synchronize the exercises/submissions.

- If the server is unreachable, the client stores the actions in local storage and synchronize them when the client is connected again.
- If the server is reachable, the client performs a `put` request (`/api/courses/{courseId}/exams/{examId}/student-exams/{studentExamId}/actions`)
  - if the request was successful, the client deletes the stored actions (no longer needed)
  - if the request was not successful, the actions are still available on the client and not removed

Since Artemis uses multiple instances, we need to share the collected actions and activities between them. For this purpose we use the distributed cache [Hazelcast](https://hazelcast.com/). During the exam, we don't want to store the actions in the database to avoid any conflicts (e.g. performance issues, ...). Therefore we save the activities and actions in the cache during the exam (and save them in the database after the exam -> future PR). 

Some comments to the current progress:

- The collection of the actions on the client needs some fine-tuning (this will be done in an additional PR)

### Steps for Testing

Testing this pull request is a little more complex.

Prerequisites:
- 1 Admin
- 1 or multiple Students
- 1 Exam with multiple Exercises
- Test Server + Test server with multiple Artemis instances (test both)

1. Log in to Artemis
2. Create an exam as admin
3. Enable the exam monitoring (Tick the `Enable Exam Monitoring` checkbox)
4. Create multiple exercises (regardless of the type of exercise)
5. Register one or multiple Students for the exam
6. Create the student exams (Two options -> need to be tested in separate exams with monitoring enabled)
a) Increase the working time of the student
b) Keep the working time as it is -> the exam end is the same as the student exam end
7. Participate as a student
8. Switch to the admin/instructor
9. Verify stored actions in Cache

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| exam-monitoring.service.ts | 85.71% | 100% |
| ExamMonitoringScheduleService.java | 62% | 91% |
| ExamCache.java | n/a | 62% |
| ExamMonitoringDistributedCache.java | 100% | 97% |
| ExamActivityResource.java | 100% | 100% | 
| ExamActivitySaveTask.java | n/a | 100% |
| SavedExerciseAction.java | n/a | 100% |
| StartedExamAction.java | n/a | 100% |
| SwitchedExerciseAction.java | n/a | 100% |
| ConnectionUpdatedAction.java | n/a | 100% |
| HandedInEarlyAction.java | n/a | 100% |
| ContinuedAfterHandedInEarlyAction.java | n/a | 100% |
| EndedExamAction.java | n/a | 100% |

### Screenshots

<details>
<summary>Screenshots</summary>

image
</details>

### CODE REVIEW

File: src/main/webapp/app/exam/participate/exam-participation.component.ts
Comment (bassner) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:202]:
```suggestion
            if (this.studentExam) {
```

File: src/test/javascript/spec/service/exam-monitoring.service.spec.ts
Comment (bassner) [inline src/test/javascript/spec/service/exam-monitoring.service.spec.ts:86]: Can we assert an absolute number of expected calls? Is ``syncActions`` called with parameters we could assert?

Same for the two cases below this comment.

**Comment by bassner** [comment]
Some suggestions to collect more useful data from events:

https://developer.mozilla.org/en-US/docs/Web/API/Window/blur_event - if the artemis tab loses focus
https://developer.mozilla.org/en-US/docs/Web/API/Window/focus_event - if the artemis tab regains focus
https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseleave_event - could be applied on the document to learn when the mouse leaves the artemis window
https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseenter_event - could be applied on the document to learn when the mouse re-enters the artemis window
https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API - watch the clipboard to see when something is copied or pasted

These combined could give some insights on suspicious behavior, such as "Tab lost focus" -> 3 secs pass -> "Tab regains focus" -> 200ms pass -> "Pasted from clipboard" ... Hmmm what could have happened ;D

**Review by ge65cer — COMMENTED** [review]
Works alright, I would suggest adding a throttle/debounce function before sending the actions request.
Also, the `type` attribute in the JSON payload is duplicated.

**Review by akesfeden — COMMENTED** [review]
Tested in testing session, works as expected. I can approve according to whether the suggestions in the comments above will be included in the current PR or later.

**Comment by Dominik-Weinzierl** [comment]
@akesfeden All suggestions are already solved in this PR (6eef026ed6881b0d906437e3c5ea4945ef63d6a5) (the enhancements mentioned by Patrick will be covered in another PR).

**Comment by bassner** [comment]
I think you need to look at the codacy issues, 16 are a lot :D

**Comment by Dominik-Weinzierl** [comment]
@bassner the amount of codacy issues should be reduced now :D

**Comment by Dominik-Weinzierl** [comment]
To make this PR testable for the partical course participants, I have removed the `isAdmin` check. This check will be added again before we merge this PR.

**Review by jbllhff — COMMENTED** [review]
I tested on ts2, but I get a 401 unauthorized message on postman. How can I fix this?

**Comment by Dominik-Weinzierl** [comment]
Solved via PM: Don't forget to add a valid access token to Postman ([Instruction](https://confluence.ase.in.tum.de/display/ArTEMiS/Postman)) :+1:

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:216]: Just an idea: Would it make sense to make this configurable in the future or is 30 minutes the sweet spot?

File: src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:39]:
```suggestion
     * This boolean indicates whether monitoring is enabled
```
Pretty sure you don't monitor rain 😉

File: src/main/java/de/tum/in/www1/artemis/domain/exam/monitoring/ExamAction.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/monitoring/ExamAction.java:30]: It might make sense to wrap the `@JsonSubTypes` annotation in `// @formatter:off` and `// @formatter:on`, so you can have one `JsonSubTypes.Type` per line. This would make the code more readable in my opinion.
But this is completely up to you if this makes your life easier.

File: src/main/java/de/tum/in/www1/artemis/domain/exam/monitoring/actions/HandedInEarlyAction.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/monitoring/actions/HandedInEarlyAction.java:9]: Considering that there is no additional data in this action, I think that this comment is not really accurate. Something like "indicates that the student handed in early" would make more sense imo.

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:731]: Could you maybe explain the rationale behind this a bit? Elsewhere in Artemis we simply directly access the ExerciseRepository when we need to load an exercise by its id.
Unless there is a good reason for this proxy method, I don't think you should deviate from this, as this would only lead to unnecessary methods in services.

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java:705]: Same here.

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java:820]: And two times here.
I really think this is not needed, as it already introduces a bunch of useless methods in services that are already very large (800+ lines in this case)

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/monitoring/ExamCache.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/monitoring/ExamCache.java:26]: What's up with all the `//` at the end of the lines here?

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/quiz/QuizCache.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/quiz/QuizCache.java:52]: You have a bunch of `//` without comments at the end of the lines here again.

**Review by Hialus — CHANGES_REQUESTED** [review]
Reviewed the server code.
Looks mostly good, but you should keep in mind that I cannot fully evaluate the quality of the Hazelcast code.
Considering that I've also tested this in a testing session and only found a minor issue that was fixed already I can already approve the manual testing.

An interesting feature and I'm looking forward to your first prototype

File: src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java
Comment (Dominik-Weinzierl) [reply src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:39]: :D

File: src/main/java/de/tum/in/www1/artemis/domain/exam/monitoring/ExamAction.java
Comment (Dominik-Weinzierl) [reply src/main/java/de/tum/in/www1/artemis/domain/exam/monitoring/ExamAction.java:30]: Yes, you are right. Thanks for the tip :D

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java
Comment (Dominik-Weinzierl) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:731]: Sure. Normally you should avoid working directly on the repositories as this is not really a good practice. I think we tried to do this in Artemis as well, but some reviewers were not so strict to enforce this behavior :thinking: (and if we try to enforce this behavior now, we get some problems with circular dependencies).

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/monitoring/ExamCache.java
Comment (Dominik-Weinzierl) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/monitoring/ExamCache.java:26]: I guess this was intended to enforce line breaks. But i use the `@formatter:off` instead :+1:

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java
Comment (Dominik-Weinzierl) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:731]: Solved via PM :+1:

**Review by Hialus — DISMISSED** [review]
Thanks for addressing my comments.
Code looks good now 👍🏻

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (bassner) [inline src/main/webapp/app/exam/monitor/exam-monitoring.service.ts:69]: As there seems to be no returned data nor any header access, we don't need to observe the full response.
```suggestion
    public syncActions(examActions: ExamAction[], courseId: number, examId: number, studentExamId: number): Observable<void> {
        const url = this.getResourceURL(courseId, examId) + `/student-exams/${studentExamId}/actions`;
        return this.http.put<void>(url, examActions);
    }
```

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (bassner) [inline src/main/webapp/app/exam/monitor/exam-monitoring.service.ts:52]: This should include a ``@returns`` tag

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (bassner) [inline src/main/webapp/app/exam/monitor/exam-monitoring.service.ts:64]: This should include a ``@returns`` tag

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (bassner) [inline src/main/webapp/app/exam/monitor/exam-monitoring.service.ts:54]: Could this be ``static``?

File: src/main/webapp/i18n/de/exam.json
Comment (bassner) [inline src/main/webapp/i18n/de/exam.json:318]: just a suggestion 🤷‍♂️ 
```suggestion
                    "title": "Klausurüberwachung aktivieren",
```

File: src/main/webapp/i18n/de/exam.json
Comment (bassner) [inline src/main/webapp/i18n/de/exam.json:319]:
```suggestion
                    "description": "Die Klausurüberwachung sammelt die Interaktionen der Studierenden mit der Klausur und bietet dadurch weitere Möglichkeiten für Statistiken."
```

File: src/test/javascript/spec/service/exam-monitoring.service.spec.ts
Comment (bassner) [inline src/test/javascript/spec/service/exam-monitoring.service.spec.ts:61]:
```suggestion
        expect(examActions).toHaveLength(Object.keys(ExamActionType).length);
```

File: src/test/javascript/spec/service/exam-monitoring.service.spec.ts
Comment (bassner) [inline src/test/javascript/spec/service/exam-monitoring.service.spec.ts:67]:
```suggestion
            expect(studentExam.examActivity?.examActions).toEqual([action]);
```

File: src/test/javascript/spec/service/exam-monitoring.service.spec.ts
Comment (bassner) [inline src/test/javascript/spec/service/exam-monitoring.service.spec.ts:75]:
```suggestion
        expect(examActions).toHaveLength(Object.keys(ExamActionType).length);
```

File: src/test/javascript/spec/service/exam-monitoring.service.spec.ts
Comment (bassner) [inline src/test/javascript/spec/service/exam-monitoring.service.spec.ts:86]: I think you forgot this one :D
```suggestion
        expect(spy).toHaveBeenCalledOnce();
```

**Review by bassner — CHANGES_REQUESTED** [review]
I've reviewed the client code. It looks really good and has a high quality, so I've been really nitpicky. :)

**Comment by Dominik-Weinzierl** [comment]
@bassner Thank you for your review and suggestions. All suggestions should be included :+1:

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (Dominik-Weinzierl) [reply src/main/webapp/app/exam/monitor/exam-monitoring.service.ts]: We need to observe the response in order to receive whether the request was successful or not :thinking: otherwise we don't know whether we can delete the actions on the client or not

**Review by rhubinger — APPROVED** [review]
Tested on TS1 in testing session with @Dominik-Weinzierl and others. Everything worked as expected. LGTM

**Review by mallgaier — APPROVED** [review]
Tested during testing session.
Works as expected, amazing job 👍

**Review by xlieA — APPROVED** [review]
Tested on ts1 during testing session. Everything worked well.

**Comment by Dominik-Weinzierl** [comment]
I removed the testing endpoint and added the isAdmin() check again :+1:

**Review by rhubinger — APPROVED** [review]
Reapprove PR, after the isAdmin() check was added again.

**Review by xlieA — DISMISSED** [review]
Reapproved after isAdmin() check was added again

**Review by nairaarnez — DISMISSED** [review]
Tested on TS1 in the testing session @Dominik-Weinzierl @rhubinger @xlieA and @mallgaier . Reaproving after isAdmin() check was added. Very good!

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (Dominik-Weinzierl) [reply src/main/webapp/app/exam/monitor/exam-monitoring.service.ts]: Yes i tried it (a08294dc2caca6299628181cfc059ba023673785) and reverted it again (a9d40b69ac49a0e587b79d987607c7670dca48fd) :thinking: can we check this again in an additional PR in order to not delay this :D

**Review by mallgaier — APPROVED** [review]
Re-approve

File: src/main/webapp/app/exam/monitor/exam-monitoring.service.ts
Comment (bassner) [reply src/main/webapp/app/exam/monitor/exam-monitoring.service.ts]: Yeah it's not that important :D

**Review by bassner — APPROVED** [review]
Thanks for adding the changes, client code looks good to me 👍

**Review by Hialus — APPROVED** [review]
Reapprove server code.
Again: I also tested this in a testing session and it worked well back then. Haven't retested it though.
Well done 👍🏻

**Review by mallgaier — APPROVED** [review]
Minor code and functionality changes since the last approval look good to me 👍

**Review by xlieA — DISMISSED** [review]
Reapprove after latest changes have been added.

**Review by Hialus — APPROVED** [review]
Only merge conflict fixes since the last review so this comment still holds true:
> Reapprove server code.
Again: I also tested this in a testing session and it worked well back then. Haven't retested it though.
Well done 👍🏻

**Review by rhubinger — DISMISSED** [review]
Reapprove after minor code changes.

**Review by nairaarnez — DISMISSED** [review]
Re-approved after changes

**Comment by Dominik-Weinzierl** [comment]
To keep the changes of the database to a minimum, we decided to store the actions only in the cache at the moment. We will think about the database changes in an additional PR. Retesting and reapproving is not required in my opinion :+1:

**Review by mallgaier — APPROVED** [review]
Retested after database changes, still works as expected.
Good job 👍

**Review by Hialus — APPROVED** [review]
The server code still looks good 👍🏻

**Review by xlieA — APPROVED** [review]
Reapprove after changes

**Review by akesfeden — APPROVED** [review]
I examined commits after my last test, the new code changes look good. Approving now with reference to the author's comment stating that retesting is not needed.

**Review by Dominik-Weinzierl — COMMENTED** [review]
I performed additional local tests and the actions are received by the server and stored in the cache. In addition, this also works correctly for multiple users and simultaneous interaction :+1: 

Thanks again for your highly appreciated reviews :tada:

**Review by nairaarnez — APPROVED** [review]
Re-approve after changes

**Pull request merged**
